### ISSUE TITLE
Proposal: vw-hypersearch for several passes

### ISSUE THREAD

**Issue by trufanov-nok** [JohnLangford/vowpal_wabbit#738]
Hi,
Usually my model requires several passes to converge on a train step. And usually every pass takes quite a long time. This was addressed before with `--save_resume` fix and support of `-c` in vw-hypersearch. I would like to suggest one more improvement.

To tune `-l` param i'm usually do following:
1. Find optimal `-l` value for single pass model with vw-hypersearch
2. Find optimal `--passes` value  with vw-hypersearch 
3. Find optimal `-l` and/or `--decay_learning_rate` values for `--passes` found on step 2.
4. Doublecheck that `--passes` is optimal for values found on step 3.

This might become quite time consuming if optimal `--passes` value is big.

There is another way to do that:
1. Find optimal `-l` value for single pass model with vw-hypersearch and save model with `--save_resume`
2. Initialize regressor with previous model state using `--save_resume` and find optimal `-l` value for pass 2 then save model with `--save_resume`
3. Initialize regressor with previous (pass 2) model state using `--save_resume` and find optimal `-l` value for pass 3 then save model with `--save_resume`
...
N. Initialize regressor with previous (pass n-1) model state using `--save_resume` and find optimal `-l` value for pass N then save model with `--save_resume`

If on N-th step the best result is worse than step N-1 result - stop and use sequence of {x1,x2,x3...xN-1} values for `-l` to train model for N-1 passes.

In other words, instead of `--decay_learning_rate` and initial `-l` value or optimal single `-l` for all passes I'm looking for sequence of `-l` values.

My suggestion is to add such mode to vw-hypersearch script. For ex. like that:
```
 Options:
        -p <N>      Find optimal values on each pass (max N passes)
        -t <TS>     Use <TS> as test-set file + evaluate goodness on it
...
```
```
$vw-hypersearch -t test_data -p 5 -L 0.01 0.8 vw data -c -k --passes 10 -l %
Warning: --passes value will be ignored due to specified -p flag
Pass 1:
trying 0.0172213810979818 ....................... 0.681695 (best)
trying 0.00580673521078512 ....................... 0.685324
trying 0.0337181738081717 ....................... 0.679407 (best)
trying 0.0510744775086489 ....................... 0.678049 (best)
trying 0.00610744775086489....................... 0.695324
best value: 0.0510744775086489
Pass 2:
trying 0.0172213810979818 ....................... 0.671695 (best)
trying 0.00580673521078512 ....................... 0.675324
trying 0.0337181738081717 ....................... 0.669407 (best)
trying 0.0510744775086489 ....................... 0.678049
best value: 0.0337181738081717
Pass 3:
trying 0.0172213810979818 ....................... 0.691695 (best)
trying 0.00580673521078512 ....................... 0.705324
trying 0.0010744775086489 ....................... 0.708049
best value: 0.0172213810979818
Optimal passes number: 2
done
```

To achieve this the script shall automatically add `-i/-f step_n.model` to the command line and may use modelname_step_n.model if user explicitly specified -f modelname.model. In case `-i` was explicitly specified it shall be used only on step 1. Generated model files shouldn't be removed after script ends so user can reuse them but may be overwritten.

What do you guys think? Does anyone else think it's useful?

**Comment by arielf** [comment]
Hi Alex,

Sorry for being slow responding to this.  `vw-hypersearch` is already a bit too complex (keeps too much global state during the search due to log-space support `-L` and other recent additions)

So I'm a bit averse to adding specific-case complexity, so if we go this route, I much prefer to do it  in a cleaner and more generic way:

Rather than hardwiring support anything specific option for `--passes` or `-l ...` I think it would be nicer and much more powerful, to support a general 2-dimensional (via `2%`) or even 3 dimensional (via `3%`) grid search by supporting a new syntax like `%<from>,<to>[,<step>]`, `2%<from>,<to>[,<step>]`, `3%<from>,<to>[,<step>]` for the place holder params.   If <step> is specified, it would imply testing in order (rather than via golden-ratio search) on all the points of the grid in a:
    `for (i = from;  i <= to; i += step) {
          # use parameter i instead of place-holder % (or <N>%)
     }`

kind of loop.

We may need to add some early termination syntax to this scheme as well.

There may be a better/cleaner syntax, this is just what occurred to me after reading the issue description.

**Comment by trufanov-nok** [comment]
Hi Martin,
Sorry for a delayed reply.
That would be a great change! Can't say that I thought a lot about proposed parameters format. But how about following:
We have currently such format: `vw-hypersearch -L a b vw_command` and ``vw-hypersearch a b vw_command`.
Let's consider this 1 dimensional case and just add step support to it: 
`vw-hypersearch -L a b vw_command` -> log-space search
`vw-hypersearch -L a b vw_command` -> golden-section search
`vw-hypersearch -L a b c vw_command` -> c is ignored and warning displayed as -L doesn't support it
`vw-hypersearch a b c vw_command` -> grid search with step `c`
you can use 0 value for `c` as default value and fallback to the golden-section search if `c == 0`. Not sure how this might be done in Perl.

And then we can add multidimensional support:
Let's use just `%` as a placeholders and address them with `-1 a b c`, '-2 a b`, `-L3 a b` args where 1,2,3 - positions of % in the command line.
Let's consider old format as default param used for all placeholders in case their params are not explicitly setted by `-n` or `-Ln. For example:

`vw-hypersearch -L a b -2 c d e vw -l % --ngram %` -> log-space search -l + grid search ngram
`vw-hypersearch -L a b -2 c d e vw -l % --ngram % --initial_t %` -> log-space search -l and -initial_t + grid search ngram
`vw-hypersearch -1 a b vw -l %` -> golden-section search -l
`vw-hypersearch a b c -L3 d f vw -i-nitial_t % --ngram %` -l % -> grid search initial_t and ngram + log-space search -l
`vw-hypersearch -L2 a b vw_command --rank % -l %` -> warning as no params are specified for --rank

My intention is to keep format as close to currently used as possible, let user specify only args that really different (by using defaults) and keep support of log-space search for inner loops. What do you think? Is it much harder to implement?

**Comment by arielf** [comment]
Hi Alex,

* Note that the current `vw-hypersearch` already supports a 3rd parameter `c` as accuracy/tolerance so proposed syntax won't work.
* There are a few typos in the above, e.g. golden and log-space search above both have `-L` option, `-i-initial_t` and more. Could you fix?
* I much prefer to have the search space and range to be tied to the `%` because they literally describe it.

I feel the cleanest/best design is to support placeholders/wild-cards that are backward compatible by allowing optional suffix-extensions to the `%` placeholder[s]:
 
     %[L][from,to]
    'L' or 'l' would mean log-space search, when unspecified it would mean golden-section search
    'from', 'to' would be the low-end, high-end of the search range

This is completely agnostic to how many `%`s we have in the line and it is much more natural to specify a space and range in the place-holder and near the flag that refers to it.  The `%` char implies a loop on the range.

The current code has a lot of global-state (global variables shared by functions) in it.  In order to make this work it will first have to be cleaned-up and re-factored with all the needed params passed to each function, to eliminate side effects.  We will also need to change the current global single min to be an array or hash of distinguishable line-search states.

**Comment by JohnLangford** [comment]
I don't have much to add, except that hyper parameter searches often do 
not take into account computational cost (and should).  Alex's original 
suggestion about using --save_resume seems like it has some real merit 
in this fashion.  An alternative approach is to use --save_per_pass and 
evaluate each model for the purpose of early stopping.

-John

**Comment by JohnLangford** [comment]
I'm closing this for now.  I still believe the hyperparameter search ideas here could be quite useful.
